bigtable/internal/testproxy/proxy.go

```diff
@@ -0,0 +1,976 @@
+	}
+	s.clientsLock.Unlock()
+
+	opts, err := getCredentialsOptions(req)
+	if err != nil {
+		return nil, stat.Error(codes.Unauthenticated,
+			fmt.Sprintf("%s: failed to get credentials: %v", logLabel, err))
+	}
+
+	conn, err := grpc.Dial(req.DataTarget, opts...)
+	if err != nil {
+		return nil, stat.Error(codes.Unknown, fmt.Sprintf("%s: failed to create connection: %v", logLabel, err))
+	}
+
+	localCtx, cancel := context.WithCancel(context.Background())
+	c, err := bigtable.NewClient(localCtx, req.ProjectId, req.InstanceId, option.WithGRPCConn(conn))
+	if err != nil {
+		cancel()
+		return nil, stat.Error(codes.Internal,
+			fmt.Sprintf("%s: failed to create client: %v", logLabel, err))
+	}
+
+	if s.clientIDs == nil {
+		s.clientIDs = make(map[string]testClient)
+	}
+
+	s.clientsLock.Lock()
+	s.clientIDs[req.ClientId] = testClient{
+		c: c,
+		cancels: []context.CancelFunc{
+			cancel,
+		},
+		appProfileID:        req.AppProfileId,
+		perOperationTimeout: req.PerOperationTimeout,
+	}
+	s.clientsLock.Unlock()
+
+	return &pb.CreateClientResponse{}, nil
+}
+
+// RemoveClient responds to the RemoveClient RPC. This method removes an
+// existing client from the goTestProxyServer
+func (s *goTestProxyServer) RemoveClient(ctx context.Context, req *pb.RemoveClientRequest) (*pb.RemoveClientResponse, error) {
+	clientID := req.ClientId
+	doCancelAll := req.CancelAll
+
+	s.clientsLock.Lock()
+	btc, exists := s.clientIDs[clientID]
+	if !exists {
+		return nil, stat.Error(codes.InvalidArgument,
+			fmt.Sprintf("%s: ClientID does not exist", logLabel))
+	}
+	defer s.clientsLock.Unlock()
+
+	if doCancelAll {
+		btc.cancelAll()
+	}
+	btc.c.Close()
+	delete(s.clientIDs, clientID)
+
+	resp := &pb.RemoveClientResponse{}
+	return resp, nil
+}
+
+// ReadRow responds to the ReadRow RPC. This method gets all of the column
+// data for a single row in the Table.
+func (s *goTestProxyServer) ReadRow(ctx context.Context, req *pb.ReadRowRequest) (*pb.RowResult, error) {
+	s.clientsLock.Lock()
+	btc, exists := s.clientIDs[req.ClientId]
+	if !exists {
+		return nil, stat.Error(codes.InvalidArgument,
+			fmt.Sprintf("%s: ClientID does not exist", logLabel))
+	}
+
+	tName := req.TableName
+	t := btc.c.Open(tName)
+
+	ctx = btc.addCancelFunction(ctx)
+
+	r, err := t.ReadRow(ctx, req.RowKey)
+	if err != nil {
+		return nil, err
+	}
+	s.clientsLock.Unlock()
+
+	if r == nil {
+		return &pb.RowResult{
+			Status: &status.Status{
+				Code: int32(codes.OK),
+			},
+			Row: &btpb.Row{},
+		}, nil
+	}
+
+	pbRow, err := rowToProto(r)
+	if err != nil {
+		return nil, err
+	}
+
+	res := &pb.RowResult{
+		Status: &status.Status{
+			Code: int32(codes.OK),
+		},
+		Row: pbRow,
+	}
+
+	return res, nil
+}
+
+// ReadRows responds to the ReadRows RPC. This method gets all of the column
+// data for a set of rows, a range of rows, or the entire table.
+func (s *goTestProxyServer) ReadRows(ctx context.Context, req *pb.ReadRowsRequest) (*pb.RowsResult, error) {
+	s.clientsLock.Lock()
```

Comment:
There's a missing unlock here that will leave the lock checked out on line 639. Best to unlock with defer.

Reply:
The unlock is missing, it's down below. I was trying to be mindful of the Big O time in these functions (but everything after the `Unlock()` happens in constant time).

Reply:
Done.